### PR TITLE
Mac CI

### PR DESCRIPTION
**Pull request by sandeepd-nv**

### CODE REVIEW

File: .github/workflows/gh-build-and-test.yml [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+            test-scope: unit
+            runner: linux-amd64-2gpu
+            enabled: ${{ inputs.platform == 'linux' }}
+
+          - name: mypy
+            test-scope: mypy
+            runner: linux-amd64-gpu-v100-latest-1
+            enabled: ${{ inputs.platform == 'linux' }}
+
+          # - name: mypy (OS X)
```
Comment (m3vaz) [inline .github/workflows/gh-build-and-test.yml:55]: Do we have someone who's following up with the things mypy is flagging?

File: conda/conda-build/meta.yaml [hunk 2]
```diff
@@ -114,8 +115,10 @@ requirements:
     - rust
     - ninja
     - cmake {{ cmake_version }}
-    - {{ compiler('c') }} =11.2
-    - {{ compiler('cxx') }} =11.2
+    - {{ compiler('c') }} =11.2   # [linux]
+    - {{ compiler('cxx') }} =11.2 # [linux]
+    - {{ compiler('c') }}   # [osx]
```
Comment (m3vaz) [inline conda/conda-build/meta.yaml:120]: Can you align the selectors so `# [osx]` is under `# [linux]`

File: continuous_integration/scripts/build-legate [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+    # Navigate to the parent of the parent of SCRIPT_DIR, then get the full path
+    REPO_DIR="$(cd "$SCRIPT_DIR/../.." && pwd)"
+
+    export REPO_DIR
+
+    export PATH="${PATH}:${REPO_DIR}/continuous_integration/scripts"
+    export ARTIFACTS_DIR="${REPO_DIR}/.artifacts"
+}
+
+# export CONDA_BUILD_THREADS=1
```
Comment (m3vaz) [inline continuous_integration/scripts/build-legate:18]: debugging?

File: continuous_integration/scripts/build-legate [hunk 4]
```diff
@@ -0,0 +1,75 @@
...
+}
+
+copy_ci_artifacts() {
+    echo Copying CI artifacts
+
+    cp -r /tmp/out          "$ARTIFACTS_DIR"
+    cp -r /tmp/conda-build  "$ARTIFACTS_DIR"
+}
+
+# NOTE: This script must be run from the project's root folder
```
Comment (m3vaz) [inline continuous_integration/scripts/build-legate:47]: Not necessarily something that needs to be done as part of the PR, but if the script is in a known place, we can use the location of the script to get back to the root folder.

File: continuous_integration/scripts/build-legate-cpp [hunk 5]
```diff
@@ -6,7 +6,11 @@ build_legate_cpp() {
     # Build + package legate.core C++ libs
     local cmake_args=(${CMAKE_ARGS:-});
     cmake_args+=(-DBUILD_SHARED_LIBS=ON);
-    cmake_args+=(-DBUILD_MARCH=${BUILD_MARCH:-haswell});
+
+    if [[ ! "$(uname)" == "Darwin" ]]; then
```
Comment (m3vaz) [inline continuous_integration/scripts/build-legate-cpp:10]: `A != B` instead of `A == B`?

File: continuous_integration/scripts/setup-utils [hunk 6]
```diff
@@ -0,0 +1,268 @@
...
+        git config --global --add user.name "anon"
+    else
+        echo "Note: git user.name is already set to $current_name"
+    fi
+
+    # Fix "fatal: detected dubious ownership in repository at '/tmp/legate.core'"
+    # during local builds.
+    git config --global --add safe.directory "$REPO_DIR"
+}
+
```
Comment (m3vaz) [inline continuous_integration/scripts/setup-utils:187]: Can you add a seperator line here, since everything above is OS-specific helper functions, everything below is a generalized function?

**Review by m3vaz — CHANGES_REQUESTED** [review]
Left a few comments, but nothing major. In principle +1

File: conda/conda-build/meta.yaml (hunk 2 above)
Comment (sandeepd-nv) [reply conda/conda-build/meta.yaml]: Done.

File: continuous_integration/scripts/build-legate (hunk 3 above)
Comment (sandeepd-nv) [reply continuous_integration/scripts/build-legate]: Removed.

File: continuous_integration/scripts/build-legate [hunk 7]
```diff
@@ -0,0 +1,75 @@
...
+# export JOBS=1
+
+build_legate_ci() {
+    build-legate-cpp;
+    build-legate-wheel;
+    build-legate-conda;
+}
+
+build_legate_release() {
+    # export NVCC_PREPEND_FLAGS='--verbose'
```
[the remark this answers is not in the document]
Comment (sandeepd-nv) [reply continuous_integration/scripts/build-legate]: Removed.

File: continuous_integration/scripts/build-legate (hunk 4 above)
Comment (sandeepd-nv) [reply continuous_integration/scripts/build-legate]: > if the script is in a known place, we can use the location of the script to get back to the root folder.
Already doing this, just forgot to remove the comment. The comment has been removed now.

File: continuous_integration/scripts/build-legate-cpp (hunk 5 above)
Comment (sandeepd-nv) [reply continuous_integration/scripts/build-legate-cpp]: Done.

File: continuous_integration/scripts/build-legate-cpp [hunk 8]
```diff
@@ -6,7 +6,11 @@ build_legate_cpp() {
...
+        cmake_args+=(-DBUILD_MARCH=${BUILD_MARCH:-x86-64});
```
[the remark this answers is not in the document]
Comment (sandeepd-nv) [reply continuous_integration/scripts/build-legate-cpp]: Removed the potentially incorrect default value from this line. Now BUILD_MARCH will be set in build-legate.

File: continuous_integration/scripts/conda-utils [hunk 9]
```diff
@@ -0,0 +1,99 @@
...
+
+    return 0;
+}
+
+get_yaml_and_make_conda_env() {
+    set -e;
+
+    local yaml_file="";
+
+    # if ! find_yaml_file; then
```
[the remark this answers is not in the document]
Comment (sandeepd-nv) [reply continuous_integration/scripts/conda-utils]: Removed.

File: continuous_integration/scripts/setup-utils (hunk 6 above)
Comment (sandeepd-nv) [reply continuous_integration/scripts/setup-utils:185]: Done.

File: .github/workflows/gh-build-and-test.yml (hunk 1 above)
Comment (sandeepd-nv) [reply .github/workflows/gh-build-and-test.yml:55]: I have asked @manopapad to help with this.

File: .github/workflows/gh-build.yml [hunk 10]
```diff
@@ -54,57 +60,34 @@ jobs:
...
+            -e AWS_REGION \
+            -e AWS_SESSION_TOKEN \
+            -e AWS_ACCESS_KEY_ID \
+            -e AWS_SECRET_ACCESS_KEY \
+            -e GITHUB_TOKEN \
+            -e USE_CUDA \
+            -v "$(pwd):$(pwd)" \
+            -v "$ARTIFACTS_DIR:$(pwd)/.artifacts" \
+            --rm "condaforge/miniforge3:latest" \
+            /bin/bash -c "cd $(pwd) && continuous_integration/scripts/build-legate ${{ inputs.build-type}}"
```
Comment (marcinz) [inline .github/workflows/gh-build.yml:78]: Do we need to run docker directly? Isn't there a way to use the `container` feature of the github runner? It seems that we should run the same `build-legate` script, but either run it in a container or not, right?

File: .github/workflows/gh-test.yml [hunk 11]
```diff
@@ -18,24 +18,84 @@ on:
...
+        type: boolean
+
+env:
+  ARTIFACT_NAME: "${{ inputs.repos-name }}-${{ inputs.build-target }}-${{ inputs.sha }}"
+  ARTIFACTS_DIR: "${{ github.workspace }}/.artifacts"
 
 jobs:
-  test:
-    if: github.repository_owner == 'nv-legate'
+  test-in-container:
```
Comment (marcinz) [inline .github/workflows/gh-test.yml:33]: I guess the same comment here. It seems that we are repeating pretty much the same steps, with the major difference being running in a container or not. Wouldn't we want the steps in a separate reusable workflow, and then we would just invoke the workflow from a container or without a container?

File: .github/workflows/gh-build.yml (hunk 10 above)
Comment (sandeepd-nv) [reply .github/workflows/gh-build.yml]: AFAIK the container feature cannot be conditionally switched off (say using if). So we will need to ...
1. Convert gh-build.yml in to a local action.
2. Create two different reusable workflows, one with container and one without and call the above local action from these workflows. 

The reusable part part come out to only 3 steps: checkout, build and upload. IMO the reusability is minimal (due to the limitations of the Github workflow markup language) and not worth the effort. Especially since the github container directive does pretty much  the same thing.

File: .github/workflows/gh-test.yml (hunk 11 above)
Comment (sandeepd-nv) [reply .github/workflows/gh-test.yml:33]: See above comment.

File: .github/workflows/gh-build.yml [hunk 12]
```diff
@@ -36,15 +38,19 @@ jobs:
...
+    defaults:
+      run:
+        shell: bash
+
     steps:
       - name: Checkout ${{ inputs.repos-name }} (= this repo)
         uses: actions/checkout@v3
         with:
           fetch-depth: 0
+          persist-credentials: false
```
[the remark this answers is not in the document]
Comment (sandeepd-nv) [reply .github/workflows/gh-build.yml:53]: Not sure what this does. Picked this up from other nv-legate repositories.

**Review by m3vaz — APPROVED** [review]

File: .github/workflows/gh-build.yml (hunk 10 above)
Comment (marcinz) [reply .github/workflows/gh-build.yml:78]: I think it may make sense to do that in the long run. Basically, creating one workflow that does the work so to say, and then have different call sites for it. Right now, we have dispatch on `use-container` anyway, so I think we actually don't add anything if we do the dispatch in a different place. Yes, things would have to move around a bit, but the benefit would be a factored out build workflow that can be called from different contexts (e.g., different driver versions, different setups, etc.).

File: .github/workflows/gh-build-and-test.yml (hunk 1 above)
Comment (sandeepd-nv) [reply .github/workflows/gh-build-and-test.yml]: It's fixed now.

File: .github/workflows/ci-gh.yml [hunk 13]
```diff
@@ -0,0 +1,35 @@
...
+  push:
+    branches:
+      - "pull-request/[0-9]+"
+      - "branch-*"
+
+jobs:
+  build-and-test:
+    strategy:
+      fail-fast: false
+      matrix:
```
Comment (marcinz) [inline .github/workflows/ci-gh.yml:17]: I think we can move container into this matrix, and then use it in the top level in gh-build-and-test. I think that should work, as everything else either runs in a container or not. But we could do that in a separate PR.

File: continuous_integration/scripts/build-legate-wheel [hunk 14]
```diff
@@ -21,15 +21,17 @@ build_legate_wheel() {
 
     local cmake_args=(${CMAKE_ARGS:-});
     cmake_args+=("-DFIND_LEGATE_CORE_CPP=ON");
-    cmake_args+=("-Dlegate_core_ROOT=$HOME/legate/build");
+    cmake_args+=("-Dlegate_core_ROOT=$REPO_DIR/legate/build");
```
Comment (sandeepd-nv) [inline continuous_integration/scripts/build-legate-wheel:24]: This should have been $REPO_DIR/build.
